### PR TITLE
[KYUUBI #1960] Skip creating sparksession and starting engine if the max initialization time exceeds

### LINKED ISSUE #1960: [Improvement] Skip creating sparksession and starting engine if the max initialization time exceeds
### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### Search before asking

- [X] I have searched in the [issues](https://github.com/apache/incubator-kyuubi/issues?q=is%3Aissue) and found no similar issues.

### What would you like to be improved?

When a resource manager such as YARN is in high traffic, many engines will be pending and fail to get initialized in time. But sometimes, the pending engine is still waiting for YARN to accept it or run it. In the meanwhile, it already becomes orphaned and may get a re-submit at the server-side. This may become an endless loop.

We should avoid these kinds of engines be running.

### How should we improve?

One way to fix this is to add a timestamp that indicates the submit time of an engine at the serverside, then before we actually create the SparkSession instance, we check whether the `current time - submit time` already exceeds the maximum engine initialization time. If true, we skip; otherwise, we go the current way

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by wForget**
### _Why are the changes needed?_

close #1960 

### _How was this patch tested?_
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [X] Add screenshots for manual tests if appropriate

Test steps:
1. Start another application to fill the queue.
2. Modify the `stop-application.sh` so it doesn't work.
3. Submit a Kyuubi request.
4. After waiting for timeout, stop the application in step 1.

We can get the following results:
1. The kyuubi request exits after timeout.
2. After obtaining resources, the orphaned engine app fails to execute and exits immediately.
image
image

- [ ] [Run test](https://kyuubi.apache.org/docs/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala [hunk 1]
```diff
@@ -168,15 +169,23 @@ object SparkSQLEngine extends Logging {
...
     SignalRegister.registerLogger(logger)
-    var spark: SparkSession = null
-    try {
-      spark = createSpark()
+    val submitTime = kyuubiConf.getOption(KYUUBI_ENGINE_SUBMIT_TIME_KEY) match {
+      case Some(t) => t.toLong
+      case _ => 0L
+    }
+    val initTimeout = kyuubiConf.get(ENGINE_INIT_TIMEOUT)
+    if (System.currentTimeMillis() - submitTime > initTimeout) {
```
Comment (wForget) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala:177]: If we do nothing here, it will exit abnormally. Is that ok?
cc @yaooqinn @turboFei

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala [hunk 2]
```diff
@@ -185,12 +194,13 @@ object SparkSQLEngine extends Logging {
...
-      case t: Throwable => error(s"Failed to instantiate SparkSession: ${t.getMessage}", t)
-    } finally {
-      if (spark != null) {
-        spark.stop()
+        }
+      } catch {
+        case t: Throwable => error(s"Failed to instantiate SparkSession: ${t.getMessage}", t)
+      } finally {
+        if (spark != null) {
+          spark.stop()
```
Comment (SteNicholas) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala:202]: The `spark` could be set to null.

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiReservedKeys.scala [hunk 3]
```diff
@@ -20,4 +20,5 @@ package org.apache.kyuubi.config
 object KyuubiReservedKeys {
   final val KYUUBI_SESSION_USER_KEY = "kyuubi.session.user"
   final val KYUUBI_STATEMENT_ID_KEY = "kyuubi.statement.id"
+  final val KYUUBI_ENGINE_SUBMIT_TIME_KEY = "kyuubi.engine.submitTime"
```
Comment (SteNicholas) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiReservedKeys.scala:23]:
```suggestion
  final val KYUUBI_ENGINE_SUBMIT_TIME_KEY = "kyuubi.engine.submit.time"
```

**Review by SteNicholas — COMMENTED** [review]
@wForget, thanks for your contribution. I have left the minor comments for this pull request.

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala (hunk 1 above)
Comment (yaooqinn) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala:177]: >  it will exit abnormally.

What you mean here is that picture you show in the PR desc?

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala [hunk 4]
```diff
@@ -168,15 +169,23 @@ object SparkSQLEngine extends Logging {
 
   def main(args: Array[String]): Unit = {
     SignalRegister.registerLogger(logger)
-    var spark: SparkSession = null
-    try {
-      spark = createSpark()
```
Comment (yaooqinn) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala:173]: we shall break `createSpark` to `setupConf` and ` createSpark` and call `setupConf` before we use `kyuubiConf`

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala (hunk 1 above)
Comment (yaooqinn) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala:177]: What if we through exception here?

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala (hunk 4 above)
Comment (wForget) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala:173]: > we shall break `createSpark` to `setupConf` and ` createSpark` and call `setupConf` before we use `kyuubiConf`

Yes, i will fix it.

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala (hunk 2 above)
Comment (wForget) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala:202]: > The `spark` could be set to null.

Sorry, I don't understand what this needs to do.

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala [hunk 5]
```diff
@@ -168,15 +169,23 @@ object SparkSQLEngine extends Logging {
...
+      warn("The engine will exit immediately, due to exceeding the maximum initialization time.")
+    } else {
+      var spark: SparkSession = null
       try {
-        startEngine(spark)
-        // blocking main thread
-        countDownLatch.await()
-      } catch {
-        case e: KyuubiException => currentEngine match {
+        spark = createSpark()
```
Comment (ulysses-you) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala:182]: the indention seems not correct ?

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/EngineRef.scala [hunk 6]
```diff
@@ -178,6 +179,7 @@ private[kyuubi] class EngineRef(
 
     conf.set(HA_ZK_NAMESPACE, engineSpace)
     conf.set(HA_ZK_ENGINE_REF_ID, engineRefId)
+    conf.set(KYUUBI_ENGINE_SUBMIT_TIME_KEY, String.valueOf(System.currentTimeMillis()))
```
Comment (ulysses-you) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/engine/EngineRef.scala:182]: can we pull out the start time of `ENGINE_INIT_TIMEOUT` ? we should keep `engine submit time` and `the timeout chekc of start time` consistent.

Comment (wForget) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/EngineRef.scala]: done

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala (hunk 5 above)
Comment (wForget) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala:187]: done

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiReservedKeys.scala (hunk 3 above)
Comment (wForget) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiReservedKeys.scala]: done

**Comment by wForget** [comment]
```
- Scheduler pool *** FAILED ***
  The code passed to eventually never returned normally. Attempted 660 times over 10.008045327 seconds. Last failure message: 0 was not greater than or equal to 1000. (SchedulerPoolSuite.scala:99)
```
The failed test doesn't seem to be caused by this pr.

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala [hunk 7]
```diff
@@ -168,29 +173,39 @@ object SparkSQLEngine extends Logging {
...
-    try {
-      spark = createSpark()
+    setupConf()
+    val submitTime = kyuubiConf.getOption(KYUUBI_ENGINE_SUBMIT_TIME_KEY) match {
+      case Some(t) => t.toLong
+      case _ => System.currentTimeMillis()
+    }
+    val initTimeout = kyuubiConf.get(ENGINE_INIT_TIMEOUT)
+    if (System.currentTimeMillis() - submitTime > initTimeout) {
+      throw new KyuubiException(s"The maximum initialization time exceeded.")
```
Comment (yaooqinn) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala:183]: let's enrich the error msg here, mention
- The total time used
- the time submitted
- the ENGINE_INIT_TIMEOUT conf and its limited value

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala [hunk 8]
```diff
@@ -70,16 +71,19 @@ case class SparkSQLEngine(spark: SparkSession) extends Serverable("SparkSQLEngin
...
 
+  val sparkConf: SparkConf = new SparkConf()
+
   val kyuubiConf: KyuubiConf = KyuubiConf()
 
   var currentEngine: Option[SparkSQLEngine] = None
 
   private val user = currentUser
 
+  val appName = s"kyuubi_${user}_spark_${Instant.now}"
```
Comment (yaooqinn) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala:82]: unnecessary change

Comment (wForget) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala:82]: > unnecessary change

`appName` is used in both `setupConf` method and `createSpark` method.

Comment (yaooqinn) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala:82]: that‘s true. but the appName is not right here. for a complete pipeline, the appName is generated by the Kyuubi server or set by users.

I suggest that we simply name the groupId in `createSpark` as "engine initializing queries "

Comment (wForget) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala:82]: > that‘s true. but the appName is not right here. for a complete pipeline, the appName is generated by the Kyuubi server or set by users.
> 
> I suggest that we simply name the groupId in `createSpark` as "engine initializing queries "

Got it, i will do that.

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/EngineRef.scala [hunk 9]
```diff
@@ -201,9 +202,10 @@ private[kyuubi] class EngineRef(
 
     MetricsSystem.tracing(_.incCount(ENGINE_TOTAL))
     try {
+      val started = System.currentTimeMillis()
+      conf.set(KYUUBI_ENGINE_SUBMIT_TIME_KEY, String.valueOf(started))
```
Comment (ulysses-you) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/engine/EngineRef.scala:206]: I prefer move these two line to line 182

Comment (wForget) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/EngineRef.scala]: > I prefer move these two line to line 182

done

**Review by yaooqinn — APPROVED** [review]

**Review by ulysses-you — APPROVED** [review]
thanks for the fix, Pending CI..

ah, seems the style is wrong. Please also fix the code style issue.

**Comment by wForget** [comment]
```
SparkHudiOperationSuite:
2022-03-01 11:42:18,195 ScalaTest-main-running-DiscoverySuite ERROR Filters contains invalid attributes "onMatch", "onMismatch"
2022-03-01 11:42:18,199 ScalaTest-main-running-DiscoverySuite ERROR Filters contains invalid attributes "onMatch", "onMismatch"
- get catalogs
- get schemas
2022-03-01 11:42:28,757 SparkSQLSessionManager-exec-pool: Thread-1794 ERROR Filters contains invalid attributes "onMatch", "onMismatch"
2022-03-01 11:42:28,770 SparkSQLSessionManager-exec-pool: Thread-1794 ERROR Filters contains invalid attributes "onMatch", "onMismatch"
- get tables *** FAILED ***
  rs1.next() was false (HudiMetadataTests.scala:85)
2022-03-01 11:42:53,888 SparkSQLSessionManager-exec-pool: Thread-1857 ERROR Filters contains invalid attributes "onMatch", "onMismatch"
2022-03-01 11:42:53,892 SparkSQLSessionManager-exec-pool: Thread-1857 ERROR Filters contains invalid attributes "onMatch", "onMismatch"
- get columns type *** FAILED ***
  0 did not equal 15 all columns should have been verified (HudiMetadataTests.scala:186)
```
The failed test doesn't seem to be caused by this pr.

**Comment by ulysses-you** [comment]
thanks, merging to master
